src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+                this.mainMenu.open(evt);
+                if (this.items && evt.type === 'enter') {
+                    this.items[0].el.focus();
+                } else {
+                    // Don't show tooltip if auto-focusing for navigation's sake.
+                    this.categoryButton.tooltip.suppress = true;
+                    this.categoryButton.element().focus();
+                }
+            }
+        } else {
+            this.el.parentNode.classList.add('jw-settings-open');
+            this.trigger('menuVisibility', { visible: true, evt });
+            document.addEventListener('click', this.onDocumentClick);
+        }
+        this.visible = true;
+        this.el.setAttribute('aria-expanded', 'true');        
+    }
+    close(evt) {
+        // if visible, return
+        if (!this.visible) {
+            return;
+        }
+        if (this.isSubmenu) {
+            this.el.classList.remove('jw-settings-submenu-active');
+            this.categoryButton.element().setAttribute('aria-checked', 'false');
+            this.mainMenu.openMenus = this.mainMenu.openMenus.filter(name => name !== this.name);
+            if (!this.mainMenu.openMenus.length && this.mainMenu.visible) {
+                this.mainMenu.close(evt);
+            }
+        } else {
+            this.el.parentNode.classList.remove('jw-settings-open');
+            this.trigger('menuVisibility', { visible: false, evt });
+            document.removeEventListener('click', this.onDocumentClick);
+            if (this.mainMenu.openMenus.length) {
+                this.mainMenu.closeAll();
+            }
+        }
+        this.visible = false;
+        this.el.setAttribute('aria-expanded', 'false');  
+    }
+    closeAll() {
+        this.openMenus.forEach(menuName => {
+            this.children[menuName].close();
+        });
+    }
+    toggle(evt) {
+        if (this.visible) {
+            this.close(evt);
+        } else {
+            this.open(evt);
+        }
+    }
+    onDocumentClick(e) {
+        if (!/jw-(settings|video|nextup-close|sharing-link|share-item)/.test(e.target.className)) {
+            this.close();
+        }
+    }
+    destroy() {
+        document.removeEventListener('click', this.onDocumentClick);
+        if (this.isSubmenu) {
+            this.parentMenu.topbar.removeChild(this.categoryButton.element());
+            this.categoryButton.ui.destroy();
+        } else {
+            this.ui.destroy();
+        }
+        this.visible = false;
+        if (this.el.parentNode) {
+            this.el.parentNode.removeChild(this.el);
+        }
+        
+        Object.keys(this.children).map(menuName => {
+            this.children[menuName].destroy();
+        });
+    }
+}
+
+const addGlobalMenuKeyListener = (settingsMenu) => {
+    const closeButton = settingsMenu.closeButton;
+    const settingsMenuElement = settingsMenu.el;
+    const ui = new UI(settingsMenuElement).on('keydown', function(evt) {
```

Comment:
where are we removing this listener on destroy?

Reply:
does `this.ui.destroy` do it?

Reply:
yep!